cmake/P4CUtils.cmake

```diff
@@ -0,0 +1,76 @@
+include (CheckCCompilerFlag)
+include (CheckCXXCompilerFlag)
+
+# test and add a C++ compiler option if supported
+MACRO (add_cxx_compiler_option option)
+  string (REPLACE "+" "P" escaped_option1 ${option})
+  string (REPLACE "-" "" escaped_option ${escaped_option1})
+  set (_success "_HAVE_OPTION_${escaped_option}_")
+  set (__required_flags_backup "${CMAKE_REQUIRED_FLAGS}")
+  set (CMAKE_REQUIRED_FLAGS "${CMAKE_REQUIRED_FLAGS} ${P4C_CXX_FLAGS}")
+  check_cxx_compiler_flag (${option} ${_success})
+  set (CMAKE_REQUIRED_FLAGS ${__required_flags_backup})
+  if (${_success})
+    set (P4C_CXX_FLAGS "${P4C_CXX_FLAGS} ${option}")
+  endif (${_success})
+endmacro (add_cxx_compiler_option)
+
+include (CheckCXXSourceCompiles)
+# check if a compiler supports a certain feature by compiling the program
+MACRO (check_compiler_feature option program var)
+  set (__required_flags_backup "${CMAKE_REQUIRED_FLAGS}")
+  set (CMAKE_REQUIRED_FLAGS "${CMAKE_REQUIRED_FLAGS} ${option}")
+
+  check_cxx_source_compiles ("${program}" ${var})
+
+  IF (${var} AND NOT "${option}" STREQUAL "")
+    set (P4C_CXX_FLAGS "${P4C_CXX_FLAGS} ${option}")
+  ENDIF ()
+  if (${var})
+    add_definitions (-D${var})
+  endif ()
+  set (CMAKE_REQUIRED_FLAGS ${__required_flags_backup})
+endmacro (check_compiler_feature)
+
+# Add files with the appropriate path to the list of linted files
+macro(add_cpplint_files dir filelist)
+  foreach(__f ${filelist})
+    list (APPEND CPPLINT_FILES "${dir}/${__f}")
+  endforeach(__f)
+endmacro(add_cpplint_files)
+
+# generate all the tests specified in the testsuites
+# Arguments:
+#   - tag is a label for the set of tests, for example, p4, p14_to_16, bmv2, ebpf
+#   - driver is the script that is used to run the tests and compare the results
+#   - testsuite is a list of directory patterns, e.g.: testdata/p4_16_samples/*.p4
+#   - xfail is a set of tests that are expected to fail
+#
+# The macro generates the test files in a directory prefixed by tag.
+#
+macro(p4c_add_tests tag driver testsuites xfail)
+  set(__xfail_list "${xfail}")
+  foreach(ts "${testsuites}")
+    file (GLOB __testfiles RELATIVE ${P4C_SOURCE_DIR} ${ts})
```

Comment:
Do you need to rerun cmake if you add a test? I'm not sure how smart its dependency management is in this case. (Not that it's a huge deal, since once we fully switch over to cmake we can implement this stuff in a more cmake-native way.)